Support pandas Series and DataFrame as input for tf.constant

yongtang opened the pull request:
This fix tries to address the proposal raised in #2318 to support pandas Series and DataFrame as input for tf.constant.

As was explained in #2318, Series and DataFrame has `__array__` method, which is the standard API used for indicating that an object is coercible into NumPy arrays.

This fix adds the support for objects with `__array__` to be served as input for tf.constant.

This fix fixes #2318.

yongtang committed:
Support pandas Series and DataFrame as input for tf.constant

This fix tries to address the proposal raised in 2318 to support
pandas Series and DataFrame as input for tf.constant.

As was explained in 2318, Series and DataFrame has `__array__`
method, which is the standard API used for indicating that an
object is coercible into NumPy arrays.

This fix adds the support for objects with `__array__` to be
served as input for tf.constant.

This fix fixes 2318.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

tensorflow-jenkins commented:
Can one of the admins verify this patch?

shoyer commented on tensorflow/python/framework/tensor_util_test.py:
```diff
@@ -662,6 +663,43 @@ def testShapeEquals(self):
     self.assertFalse(tensor_util.ShapeEquals(t, [4]))
 
+  def testSeries(self):
+    df = pd.Series([10, 20, 30])
```
I would suggest testing this with a dummy class that implements `__array__` rather than actually using pandas.

yongtang committed:
Add test case to support object with `__array__` as input for tf.constant

This commit adds a test case to support object with `__array__` as
input for tf.constant.

This fix is part of the PR to address 2318.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
@shoyer Thanks for the review. The PR has been updated with added test case of `testMockArray` which has the dummy class that implements `__array__`. Please take a look.

shoyer commented:
Just to be clear, I'm Googler and TensorFlow contributor, but not an owner for this part of the code. Somebody else will triage this PR and assign it for review.

martinwicke commented on tensorflow/python/framework/tensor_util_test.py:
```diff
@@ -19,6 +19,7 @@
 from __future__ import print_function
 
 import numpy as np
+import pandas as pd
```
We can't have unguarded pandas imports -- you can try to import it -- see pandas imports in contrib/learn

martinwicke replied:
Actually, please see @shoyer's comment below. Please remove pandas from this.

alextp commented on tensorflow/python/framework/tensor_util_test.py:
```diff
@@ -662,6 +663,69 @@ def testShapeEquals(self):
+    self.assertAllClose(np.array([10, 20, 30], dtype=np.int64), a)
+
+  def testDataFrame(self):
+    df = pd.DataFrame({'A': [10], 'B': [20], 'C': [30]})
```
Let's test only a generic object with the __array__ method and comment that this supports pandas.

alextp commented on tensorflow/python/framework/tensor_util_test.py:
```diff
@@ -662,6 +663,69 @@ def testShapeEquals(self):
+      def __array__(self, dtype=None):
+        return np.asarray(self.array, dtype)
+
+    ma = MockArray(np.array([10, 20, 30]))
```
Can you make these arrays smaller (maybe single-element) so the proto will fit in the test?

alextp commented on tensorflow/python/framework/tensor_util_test.py:
```diff
@@ -662,6 +663,69 @@ def testShapeEquals(self):
+
+    ma = MockArray(np.array([10, 20, 30]))
+    t = tensor_util.make_tensor_proto(ma)
+    if sys.byteorder == "big":  
```
Actually, instead of using tensor proto, which relies on big vs little endian, how about just session.run the return value of ops.convert_to_tensor and comparing with the original .__array__ result?

alextp reviewed (changes requested):
Thanks for the contribution!

I like the added support; I just have concerns about the tests which should be easy to address.

yongtang committed:
Remove pandas imports, and use convert_to_tensor, as suggested in the review.

This fix removes pandas imports, and use convert_to_tensor,
as suggested in the review.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @martinwicke @alextp for the review. The PR has been updated as suggested. Please take a look.

martinwicke commented:
Jenkins, test this please.
